File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {

        List<List<Integer>> result = new ArrayList();
        generatePaths(result, root, targetSum, new ArrayList());
        return result;
    }

    private void generatePaths(List<List<Integer>> result, TreeNode node, int target, List<Integer> list){

        if(node==null)
            return;
        list.add(node.val);
        if(node.left==null && node.right==null){
            if(target==node.val){
                result.add(new ArrayList(list));
            }
            
        }
        else{
            generatePaths(result, node.left, target-node.val, new ArrayList(list) );
            generatePaths(result, node.right, target-node.val, new ArrayList(list));
        }
        list.remove(list.size()-1);
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    vector<vector<int>> ans;
    
    void utilPathSum(TreeNode* r, int T, vector<int> temp){
        
        if(!r) return;
        temp.push_back(r->val);
        T-=r->val;
        
        if(!r->left and !r->right and T==0){
            ans.push_back(temp);
            return;
        }
        
      utilPathSum(r->left, T, temp);
      utilPathSum(r->right, T, temp);  
    }
    
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
    
        vector<int> temp;
        utilPathSum(root, targetSum, temp);
        
        return ans;
    }
};